lib/state_machine/src/types.rs

```diff
@@ -0,0 +1,135 @@
+            "ready" => Some(State::Ready),
+            "accelerate" => Some(State::Accelerate),
+            "lim_brake" => Some(State::LimBrake),
+            "friction_brake" => Some(State::FrictionBrake),
+            "stop_levitation" => Some(State::StopLevitation),
+            "stopped" => Some(State::Stopped),
+            "emergency_brake" => Some(State::EmergencyBrake),
+            "safe" => Some(State::Safe),
+            "shutdown" => Some(State::Shutdown),
+            _ => None,
+        }
+    }
+
+    pub fn transition(current_state: &State, to_state: &State) -> Option<State> {
+        let to_from_state = (current_state, to_state);
+
+        match to_from_state {
+            (State::Idle, State::Calibrate) => Some(State::Calibrate),
+            (State::Calibrate, State::Precharge) => Some(State::Precharge),
+            (State::Precharge, State::ReadyForLevitation) => Some(State::ReadyForLevitation),
+            (State::ReadyForLevitation, State::BeginLevitation) => Some(State::BeginLevitation),
+            (State::BeginLevitation, State::Levitating) => Some(State::Levitating),
+            (State::Levitating, State::Ready) => Some(State::Ready),
+            (State::Ready, State::Accelerate) => Some(State::Accelerate),
+            (State::Accelerate, State::LimBrake) => Some(State::LimBrake),
+            (State::Accelerate, State::EmergencyBrake) => Some(State::EmergencyBrake),
+            (State::LimBrake, State::FrictionBrake) => Some(State::FrictionBrake),
+            (State::FrictionBrake, State::StopLevitation) => Some(State::StopLevitation),
+            (State::StopLevitation, State::Stopped) => Some(State::Stopped),
+            (State::Stopped, State::Safe) => Some(State::Safe),
+            (State::EmergencyBrake, State::Safe) => Some(State::Safe),
+            (State::Safe, State::Shutdown) => Some(State::Shutdown),
+            _ => None,
+        }
+    }
+
+    pub fn get_macro_state(state: &State) -> MacroState {
+        match state {
+            State::Idle => MacroState::Idle,
+            State::Calibrate => MacroState::Idle,
+            State::Precharge => MacroState::Active,
+            State::ReadyForLevitation => MacroState::Active,
+            State::BeginLevitation => MacroState::Active,
+            State::Levitating => MacroState::Active,
+            State::Ready => MacroState::Active,
+            State::Accelerate => MacroState::Demo,
+            State::LimBrake => MacroState::Demo,
+            State::FrictionBrake => MacroState::Demo,
+            State::StopLevitation => MacroState::Demo,
+            State::Stopped => MacroState::Active,
+            State::EmergencyBrake => MacroState::Emergency,
+            State::Safe => MacroState::Idle,
+            State::Shutdown => MacroState::Idle,
+        }
+    }
+}
+
+pub enum MacroState {
+    Idle,
+    Active,
+    Demo,
+    Emergency,
+}
+
+impl MacroState {
+    pub fn to_string(&self) -> String<20> {
```

Comment:
very nitpick, do these need to be 20 bytes long? is that a convention somewhere?